### ISSUE TITLE
Symmetric affine int8 but appering -128 in convolution weight when using 16x8 quantization

### ISSUE THREAD

**Issue by steven0129** [alibaba/TinyNeuralNetwork#60]
Perform below code

```python
if __name__ == '__main__':

    with model_tracer():
        
        model = QfMobileNet()
        model.eval()
        model.softmax = nn.LogSoftmax(dim=1)
        
        dummy_input = torch.rand((1, 1, 135, 240))
        quantizer = QATQuantizer(
            model, dummy_input, work_dir=output_dir, config={
                'backend': "qnnpack", 
                'force_overwrite': True, 
                'asymmetric': False, 
                'per_tensor': False, 
                'rewrite_graph': True
            }
        )
        qat_model = quantizer.quantize()
        
        device = get_device()
        qat_model.to(device=device)
        
        context = DLContext()
        context.device = device
        context.train_loader = training_loader
        context.val_loader = testing_loader
        context.max_epoch = 100
        context.criterion = nn.NLLLoss()
        context.optimizer = torch.optim.Adam(qat_model.parameters(), lr=0.001)
        if isinstance(context.criterion, nn.Module):
            context.criterion = context.criterion.to(device=context.device)

        qat_model(torch.randn(1, 1, 135, 240).cuda())

        with torch.no_grad():
            qat_model.eval()
            qat_model.cpu()
            qat_model.softmax = nn.Softmax(dim=1)
            qat_model = torch.quantization.convert(qat_model)
            torch.backends.quantized.engine = 'qnnpack'
            converter = TFLiteConverter(qat_model, dummy_input, tflite_path=f'{output_dir}/qat.tflite', quantize_target_type='int16', strict_symmetric_check=True)
            converter.convert()
```

Then I got qat.tflite but -128 is at convolution weight which input name is "input.4_te_transform_1_te_transform_2" even though setting asymmetric as False (Range shoud be [-127, 127])

![image](https://user-images.githubusercontent.com/18100669/159646888-45874c51-9014-4b9a-8495-07a9ef328c33.png)

qat.tflite is compressed as [qat.zip](https://github.com/alibaba/TinyNeuralNetwork/files/8330797/qat.zip)

**Comment by peterjc123** [comment]
@steven0129 See https://github.com/alibaba/TinyNeuralNetwork/blob/main/tinynn/graph/quantization/quantizer.py#L471-L480. The quant ranges of the weight tensors are only reset for the qnnpack backend. Maybe we should apply it for the fbgemm backend as well.

**Comment by steven0129** [comment]
@peterjc123 

I set backend to "qnnpack" but not "fbgemm"

```python
quantizer = QATQuantizer(
      model, dummy_input, work_dir=output_dir, config={
          'backend': "qnnpack", 
          'force_overwrite': True, 
          'asymmetric': False, 
          'per_tensor': False, 
          'rewrite_graph': True
      }
  )
```

**Comment by peterjc123** [comment]
@steven0129 Will take a look.

**Comment by steven0129** [comment]
@peterjc123 Many thanks.

**Comment by peterjc123** [comment]
@steven0129 I have to say this is not a bug. When I changed https://github.com/alibaba/TinyNeuralNetwork/blob/main/tinynn/graph/quantization/quantizer.py#L352-L360 to the following, the quantized values of the weights are all within -127 and 127.
```py
                sym_fq = torch_q.FakeQuantize.with_args(
                    observer=torch_q.PerChannelMinMaxObserver.with_args(quant_min=-127, quant_max=127),
                    quant_min=-127,
                    quant_max=127,
                    dtype=torch.qint8,
                    qscheme=torch.per_channel_symmetric,
                    reduce_range=False,
                    ch_axis=0,
                )
```
The reason why the q-value overflows is because 
1. You are training with random data
2. You are training with only one step
3. `PerChannelMinMaxObserver` is used by default, so the scales and offsets are not calculated using the current qmin and qmax

As for QAT, you need to freeze the q-params in the late stage of training (usually in the last 1/3 of the training epochs).

**Comment by steven0129** [comment]
Hi @peterjc123 

Okay, I got it. Thanks for your information. This issue can be closed.

**Comment by steven0129** [comment]
@peterjc123 Excuse me, how to freeze the q-params in the late stage of training specifically?

**Comment by peterjc123** [comment]
Please see https://github.com/alibaba/TinyNeuralNetwork/blob/main/tinynn/util/train_util.py#L121-L127.

**Comment by steven0129** [comment]
Hi @peterjc123 
I increase code as below

```python
def run(self):
        torch.manual_seed(self.random_seed)
        for epoch in range(self.context.max_epoch):
            self.context.epoch = epoch
            self.train()

            if self.context.epoch == self.context.max_epoch // 3:
                log.info("freeze quantizer parameters")
                self.model.apply(torch.quantization.disable_observer)
            elif self.context.epoch == self.context.max_epoch // 3 * 2:
                log.info("freeze batch norm mean and variance estimates")
                self.model.apply(torch.nn.intrinsic.qat.freeze_bn_stats)

            self.test()
            torch.save(self.model.state_dict(), f'{self.output_folder}/qat_model_epoch{epoch}.pth')
```

However, I still got -128 in my model weight. Is there any ways to check whether I succeed to freeze q-params?

**Comment by peterjc123** [comment]
> Is there any ways to check whether I succeed to freeze q-params?

You can just `print(qat_model)` and then you may find the `weight_fake_quant` object in the specific layer. And if it is not updating the quantization parameters, you can see something like `observer_enabled=tensor([0], device='cuda:0')` in it.

BTW, please note the following things.
1. The training logic won't work well if you train from scratch.
2. Make sure you set a rather low learning rate.
